drivers/ds18/ds18.c

```diff
@@ -0,0 +1,232 @@
+        if (ds18_read_bit(dev, &bit) == DS18_OK) {
+            *byte |= (bit << i);
+        }
+        else {
+            return DS18_ERROR;
+        }
+    }
+
+    return DS18_OK;
+}
+
+static void ds18_write_byte(ds18_t *dev, uint8_t byte)
+{
+    for (int i = 0; i < 8; i++) {
+        ds18_write_bit(dev, byte & (0x01 << i));
+    }
+}
+
+static int ds18_reset(ds18_t *dev)
+{
+    int res;
+
+    /* Line low and sleep the reset delay */
+    ds18_low(dev);
+    xtimer_usleep(DS18_DELAY_RESET);
+
+    /* Release and wait for the presence response */
+    ds18_release(dev);
+    xtimer_usleep(DS18_DELAY_PRESENCE);
+
+    /* Check device presence */
+    res = gpio_read(dev->pin);
+
+    /* Sleep for reset delay */
+    xtimer_usleep(DS18_DELAY_RESET);
+
+    return res;
+}
+
+int ds18_trigger(ds18_t *dev)
+{
+    int res;
+
+    res = ds18_reset(dev);
+    if (res) {
+        return DS18_ERROR;
+    }
+
+    /* Please note that this command triggers a conversion on all devices
+     * connected to the bus. */
+    ds18_write_byte(dev, DS18_CMD_SKIPROM);
+    ds18_write_byte(dev, DS18_CMD_CONVERT);
+
+    return DS18_OK;
+}
+
+int ds18_read(ds18_t *dev, int16_t *temperature)
+{
+    int res;
+    uint8_t b1 = 0, b2 = 0;
+
+    DEBUG("[DS18] Reset and read scratchpad\n");
+    res = ds18_reset(dev);
+    if (res) {
+        return DS18_ERROR;
+    }
+
+    ds18_write_byte(dev, DS18_CMD_SKIPROM);
+    ds18_write_byte(dev, DS18_CMD_RSCRATCHPAD);
+
+    if (ds18_read_byte(dev, &b1) != DS18_OK) {
+        DEBUG("[DS18] Error reading temperature byte 1\n");
+        return DS18_ERROR;
+    }
+
+    DEBUG("[DS18] Received byte: 0x%02x\n", b1);
+
+    if (ds18_read_byte(dev, &b2) != DS18_OK) {
+        DEBUG("[DS18] Error reading temperature byte 2\n");
+        return DS18_ERROR;
+    }
+
+    DEBUG("[DS18] Received byte: 0x%02x\n", b2);
+
+    int32_t measurement = ((int32_t)(b2 << 8 | b1) * 625);
+    *temperature = (int16_t)(measurement / 100);
+
+    return DS18_OK;
+}
+
+int ds18_get_temperature(ds18_t *dev, int16_t *temperature)
+{
+    int res;
+
+    res = ds18_reset(dev);
+    if (res) {
+        return DS18_ERROR;
+    }
+
+    DEBUG("[DS18] Convert T\n");
+    if (ds18_trigger(dev)) {
+        return DS18_ERROR;
+    }
+
+    DEBUG("[DS18] Wait for convert T\n");
+    xtimer_usleep(DS18_DELAY_CONVERT);
+
+    return ds18_read(dev, temperature);
+}
+
+int ds18_init(ds18_t *dev, const ds18_params_t *params)
+{
+    int res;
+
+    /* Deduct the input mode from the output mode. If pull-up resistors are
+     * used for output then will be used for input as well. */
+    dev->in_mode = (params->out_mode == GPIO_OD_PU) ? GPIO_IN_PU : GPIO_IN;
+
+    /* Initialize the device and the pin */
+    dev->pin = params->pin;
+    dev->out_mode = params->out_mode;
+    res = gpio_init(dev->pin, dev->in_mode) == 0 ? DS18_OK : DS18_ERROR;
+
+#if defined(MODULE_DS18_OPTIMIZED)
+    DEBUG("[DS18] Using optimized read function");
+#endif
```

Comment:
please remove this block, I think it is rather for internal debugging...

Reply:
Removed